### ISSUE TITLE
Reorder items in the NVDA menu

### ISSUE THREAD

**Issue by Adriani90** [nvaccess/nvda#15769]
### Is your feature request related to a problem? Please describe.
Not really, it is just a thing about qualitative design and a more convenient menu structure. There is however the problem of conflicting accelerator keys again and again, especially when adding new things to the menu.

### Describe the solution you'd like
I propose following for the NVDA menu:
1. Under Prefferences sub menu, rename the "settings" item to "configuration" and make it also a submenu. The configuration sub menu should include following items:
• "settings"
• configuration profiles
• Revert to saved configuration
• Reset configuration to factory defaults
• Save configuration

Benefits: the startpage of the NVDA menu gets cleaner and easier to understand for beginners, and the items in the configuration sub menu would not really break user experience because more advanced users use rather the key commands to display them (e.g. ctrl+nvda+p for the profile dialog).

2. The dialogs "punctuation / symbol pronounciation" and "input gestures" found under prefferences sub menu should be transformed into settings categories and moved into the settings _dialog. They could be invoked via the shorcuts ctrl+nvda+i "input gestures dialog" and ctrl+nvda+y "punctuation / symbol pronounciation".

So if implemented, the start page of the NVDA menu should look like following:
• Prefferences sub menu
• Tools sub menu
• Help sum menu
• Donate
• Exit

And The preferences sub menu should contain following items:
• configuration sub menu
• Speech dictionaries sub menu

### Describe alternatives you've considered
Leave as is.

### Additional context
None

**Comment by Adriani90** [comment]
Actually the configuration profiles dialog could also be made a category in the settings dialog rather than a standalon menu item in the NVDA menu. We already have ctrl+nvda+p to open it, so it is already kind of a settings category from its nature.

**Comment by mzanm** [comment]
I disagree with putting the settings item in a configuration menu in the preferences menu. It's the most used, and putting it 3 menus deep isn't a good user experience. It's not convenient if you change settings often. There's a shortcut, but some might be used to the current layout, and this will just add rarely used features in the way of what's most commonly used.

**Comment by Adriani90** [comment]
@mzanm So can we agree that the items I proposed for the configuration sub menu could at least be moved from the start page of the NVDA menu to the preferences dialog, and the items configuration profiles, input gestures and punctuation / symbols dialog should be moved to the settings dialog and made as settings categories instead of menu items? This would be at least more consistent with general software design and it provides more structure to the users.

**Comment by CyrilleB79** [comment]
>     2. The dialogs "punctuation / symbol pronounciation" and "input gestures" found under prefferences sub menu should be transformed into settings categories and moved into the settings _dialog. They could be invoked via the shorcuts ctrl+nvda+i "input gestures dialog" and ctrl+nvda+y "punctuation / symbol pronounciation".

I strongly disagree with this point. The majority of the categories of the settings dialog is profile dependent. Adding the keyboard shortcut there would let think that they may be profile dependent. Same for the profile dialog which cannot be profile dependent at all since it is used to define them! That would be very confusing.
I acknowledge that the General category is an exception and is not profile dependent; but IMO, adding other categories that do not depend on profile in the settings dialog would just increase confusion.


>     3. The info dialog should contain following elements partially moved from the help menu to this dialog:
>        • A read only edit field with the information about NVDA
>        • License button
>        • Contributors button
>        • Go to NV Access website button
>        • Ok button

Why put here a "Go to NV Access website" button and keep other NV Access new links in the help menu? That too would be confusing.

**Comment by XLTechie** [comment]
@Adriani90 Can you break the About dialog changes (you call it Info dialog) into 
a separate issue? It's not related to configuration/settings/etc.
I support that mostly, other than what Cyrille has mentioned about the 
link, but I am undecided about the majority of the configuration changes you 
propose.

I would like to work on the one without effecting the others.

**Comment by Adriani90** [comment]
@XLTechie as soon as we have a common understanding of the design in the end I will split them.

**Comment by Adriani90** [comment]
@CyrilleB79 
> adding other categories that do not depend on profile in the settings dialog would just increase confusion.

Why that? NVDA imediately reports when you enter a settings category which is profile dependend as soon as you have a profile that is currently triggered. Where would the confusion come from?

> Why put here a "Go to NV Access website" button and keep other NV Access new links in the help menu?

In my view the wording "get help and support" is by its nature better suited in the help menu than the info dialog. The other point with the NV Access shop, I am not really sure how to handle that because in my opinion the shop is already on the website, so it should suffice to have the "go to NV Access website" in the info dialog. I understand NV Access's attempt to advertise more for their products, but in my view the help menu is also the wrong place for the NV Access shop.

The NVAccess website, license and the mentioning of contributors are already part of the info text in the info dialog, so users would expect to access those links particularly in this dialog. The get help and support or the NV Access shop are not part of the info text in the info dialog, so users would not expect them here.

**Comment by seanbudd** [comment]
I would not move the welcome dialog to preferences from help. The dialog serves to help the user get started with NVDA, by providing basic help and also configuration. I think the welcome dialog serves more to help than configure, as you can change the same settings and more in the settings dialog.

>The other point with the NV Access shop, I am not really sure how to handle that because in my opinion the shop is already on the website, so it should suffice to have the "go to NV Access website" in the info dialog. I understand NV Access's attempt to advertise more for their products, but in my view the help menu is also the wrong place for the NV Access shop.

The reason it was included is the NV Access shop sells mostly training materials and support.

**Comment by Adriani90** [comment]
I edited the description according to the common understanding so far. The part about the info dialog is now in a separate feature request.
The point remains open whether we should add a configuration submenu under prefferences sub menu or not. In my view it makes sense, because speech dictionary for example is also a sub menu so it is common to have two consecutive sub menus.

**Comment by seanbudd** [comment]
Closing this - it seems that rearranging the menu like this is unpopular

**Comment by Adriani90** [comment]
@seanbudd could you please elaborate more why do you think this is unpopular?
I mean it is obvious that at some point we will need a reorder or a redesign of the menu (I) due to various accelerator keys conflicting witheach other in the future when new things are added and (II) due to new features for interface design in WX python which bring improvements such as other types of lists etc.

With my proposal, and together with #15800 we would get rid of at least 6 accelerator keys from the Menu, which free up space for potential other menu items we don't know of yet.

cc: @michaelDCurran

**Comment by Adriani90** [comment]
I feel like this needs more feedback from the community before closing.

cc: @lukaszgo1, @josephsl, @aaclause, @bramd could you also add your opinion on this one?

**Comment by josephsl** [comment]
Hi,

A few things:

* How did we end up with current menu structure: we need to start there, otherwise the proposal may not make sense. In old days, settings panels were individual dialogs, hence they were placed inside preferences menu. With NVDA 2018.2, most of the individual settings dialogs became panels in the Settings dialog. The exceptions are speech dictionaries, punctuation/symbol pronunciation, and input gestures dialogs. This is because unlike Settings dialog, these three items use completely different ways to store settings - settings dialog options are housed inside a multi-dimensional map (config.conf) whereas speech dictionaries, symbol pronunciation, and input gestures are housed in dictionary files, symbols.dic, and gestures.ini, respectively. In theory, we could put these dialogs inside the settings dialog, but this means we need to edit the code that allows NVDA to load settings whenever you switch between panels, and that work is quite involved.
* Speech dictionaries: at first glance, speech dictionaries as a submenu may give an impression that we can create nested menus for other menu items. However, we need to remember that speech dictionaries is really one thing divided into three different scopes: at the global level (default dictionary), voice level (voice dictionary), and for experimentation (temporary dictionary). If you look at it this way, speech dictionaries is no longer viewed as a submenu, but a standalone menu iteu.
* Configuration profiles: it can be argued that configuration profiles can become part of settings dialog as profiles collect settings to be used in specific situations. However, the purpose and internals of configuration profiles makes it harder to follow the proposal. First, configuration profiles do not work at a limited view of NVDA settings like settings panels are. Configuration profiles work with a collection of NVDA settings that span settings panels; changing profiles do not change a single setting - changing profiles can change multiple settings at once (profiles treat a collection of settings as a single unit). Second, internally, configuration profiles are not just a collection of ini files - profiles are custom classes (objects at runtime) and are organized and treated differently from config.conf. If you want to look at a simplified view of how configuration profiles work, take a look at StationPlaylist add-on source code.
* Speaking of add-ons: the biggest issue I find with the proposal is that we might be forgetting the vast ecosystem of add-ons. When add-ons add menu entries, accelerator collisions are inevitable, more so for add-ons with translated menu entries. While many add-ons with settings interface do employ settings panels, some do not - some add-ons do come with settings dialogs with their own multiple settings panels.
* Localization: we might be thinking about English interface as part of this proposal. But let us continue to remind ourselves that this also affects other languages, more so when add-ons are involved.

My take is this: proposing to change NVDA menu structure changes how NVDA communicates to users about how to find things using its menu structure. Therefore, we need to think really carefully about this proposal because adopting the ideas proposed would mean we are committing to changing our own communication practices when interacting and teaching users and those outside the NVDA community. At the moment, the NVDA source code and what it communicates (at the levels of code, infrastructure, Git log, user interface, you name it) makes it a bit harder to adopt the proposal as is (see my notes above).

Postscript: I think one thing we might not have thought about (perhaps until now) is tension between who or what has authority and agency. Surely, the community has vast authority and agency as demonstrated by the proposal here. At the same time, the fact that NVDA is open-source makes it possible to view source code as having authority and agency even if it is written by humans. Thus, changing the source code, a text that communicates authority and agency, means we are creating (more towards constructing) a new reality and means of communicating with not only people, but with NVDA (as a technology) itself. Some of us may say, "Joseph, this is very deep" or "I just want to change the user interface to make it better organized." While I understand this, I believe that when dealing with software that is open-source (open to public scrutiny), how the current code came to be and its workings (and possibilities of change) should not be discounted because discounting the code and infrastructure (along with human understanding) amounts to building something without thinking about materials.

Thanks.

**Comment by CyrilleB79** [comment]
I am not in favor of changing the menu structure (due to existing material and teaching practice) unless it's to solve a real issue.

IMO, the problem of accelerator conflicting keys is not a significant enough problem. We have had such issues when adding the add-on store, but it was quite easy to fix this. And overall, in any case, if we use accelerator keys for all item, a conflict just means that the item will not be activated when the accelerator key is pressed, but it will be focused in any case alternatively with the other item having the same accelerator key. That's not a very important problem and it's not worth changing the whole menu structure just for this.

More generally, I have not heard that the menus are confusing for new users, at least from my readings on the mailing lists; I have no personal NVDA teaching experience. I have not heard either than the start page of the menu contains too much items, making it difficult to understand.

Anyway, one interesting point in the current proposal (and which could help understanding NVDA configuration) is the idea to gather all elements related to configuration in a specific sub-menu. However, I would do something a little different from what is proposed in the current PR:
The structure of NVDA menu would be the following:
- Configuration
  - Settings
  - Configuration profiles
  - Revert to saved configuration
  - Reset configuration to factory defaults
  - Save configuration
- Other preferences (or any other better name)
  - Dictionaries
  - Punctuation / symbol pronounciation
  - Input gestures
- Tools (sub-menu unchanged)
- Help (sub-menu unchanged)
- Donate
- Exit

This structure has the advantage of separating the configuration elements which are impacted by the profiles and those which are not. Indeed, it may be a point of confusion for new users to see that gestures, dictionaries or symbol description is not impacted by profile change; this would be clarified by such menu structure. Also, I avoid double nested sub-menus, except for dictionaries as today.

**Comment by Adriani90** [comment]
Thank you so far for this valuable feedback. It helps alot to understand the mechanisms behind it.

@josephsl wrote:
> In theory, we could put these dialogs inside the settings dialog, but this means we need to edit the code that allows NVDA to load settings whenever you switch between panels, and that work is quite involved.

Understood. However, I am still of the opinion that from a UX perspective it would make sense to include input gestures, profile dialog and symbols / punctuation dialog in the settings categories. We would end up with less different references when navigating new users to these different parts of NVDA. So maybe someone in the future will have the motivation to do this work which I agree is not trivial but it is possible. I am supporting lots of users comming to NVDA from other screen readers and they all struggle with multiple references in the user guide, in tutorials etc. They rather like a dialog where they can do most of the settings, including gestures, symbol pronounciation, profile management etc. So unless it is tecnically impossible, we should still let this option open for people who want to do the work on this.

> Speech dictionaries: at first glance, speech dictionaries as a submenu may give an impression that we can create nested menus for other menu items. However, we need to remember that speech dictionaries is really one thing divided into three different scopes:

I agree and thatt's why I left this part untouched in my request. This should remain as it is.

> However, the purpose and internals of configuration profiles makes it harder to follow the proposal. First, configuration profiles do not work at a limited view of NVDA settings like settings panels are. Configuration profiles work with a collection of NVDA settings that span settings panels; changing profiles do not change a single setting - changing profiles can change multiple settings at once (profiles treat a collection of settings as a single unit). Second, internally, configuration profiles are not just a collection of ini files - profiles are custom classes (objects at runtime) and are organized and treated differently from config.conf.

I understand the technical part of this and agree it is not trivial to move it, however from an UX design perspective the user uses this dialog to manage its profiles, so it is a profile management dialog which changes different settings according to the profile you choose. In my view it is logical from its functional nature to include it into the settings categories. The technical functions behind it should ofcourse remain the same.

> the biggest issue I find with the proposal is that we might be forgetting the vast ecosystem of add-ons. When add-ons add menu entries, accelerator collisions are inevitable, more so for add-ons with translated menu entries.

I am not trying to request to eliminate completely accelerator key conflicts. But with a slightly different menu structure we can minimize it at least. Especially for novice users it is really a mess to deal with multiple targets for the same accelerator key, and until people learn more efficient ways of accessing dialogs, accelerator keys are the way to go, even for advanced users, but even more for novice users.

> Localization: we might be thinking about English interface as part of this proposal. But let us continue to remind ourselves that this also affects other languages, more so when add-ons are involved.

I would let solving this issues in the responsibility of translators and addon authors, as it is now too. We will never be able to solve everything, and this was also not my intention when I did this request.

> While I understand this, I believe that when dealing with software that is open-source (open to public scrutiny), how the current code came to be and its workings (and possibilities of change) should not be discounted because discounting the code and infrastructure (along with human understanding) amounts to building something without thinking about materials.

Not sure I understand completely your point, but in fact yes, the origine of a code should not be an impediment for change. In my view changing something, if it has a positive impact, then the option should be there. I know this proposal might be trivial for advanced users, but in my experience of teaching people NVDA, it would ease the work significantly.
However, I also understand your point that it is technically complex, but this should not be the argument to close a request.

> proposing to change NVDA menu structure changes how NVDA communicates to users about how to find things using its menu structure. Therefore, we need to think really carefully about this proposal because adopting the ideas proposed would mean we are committing to changing our own communication practices when interacting and teaching users and those outside the NVDA community.

That's why I made this proposal, because with the current menu structure things often do not seem logical for new users, and I still think the changes proposed here would make it easier for new people to understand the mechanics behind certain features in NVDA. In the end this proposal does not significantly change the design of NVDA as such, it is just a restructuring which can easilby be considered in updated tutorials and user guide and what not. Best example: windows settings center and the old control pannel. Users got used to the new settings center quite easily because it added more structure and made things more understandable.

@CyrilleB79 wrote:
> I am not in favor of changing the menu structure (due to existing material and teaching practice) unless it's to solve a real issue.

In fact it is tackling more real issues as explained above. (I) the accelerator key issue and (II) the different references to different parts of NVDA which seems still to be a pain point for people still outside of the NVDA or new to the NVDA community.

> That's not a very important problem and it's not worth changing the whole menu structure just for this.

This is an advanced NVDA user perspective, which I agree. Ho wever, in practice it is not the same experience when you teach new people NVDA.

> This structure has the advantage of separating the configuration elements which are impacted by the profiles and those which are not.

Overall I agree your proposal would at least make some things better and would be at least a good start.

**Comment by lukaszgo1** [comment]
TL;DR I agree that this issue should remain closed.
Replying to some points brought up in the above discussion. Firstly I strongly agree that complexity (in terms of code changes required) should be considered last. In this particular case we should agree on a design from the UX perspective first, and then eventually start thinking about how complex it would be to implement, so that should not be argument either in favor, or against the changes proposed.
In my opinion the first  question which should be answered here (or for that matter in any other UI changes proposal) is a problem statement, i.e. "Because of X, users cannot do Y, with the changes described in Z Y becomes easier because...":. That has not been  done here, we just have a qualitative opinion from @Adriani90 that in their view this change would make things easier for beginners. Opinions like these should not be discounted by any means, but They should not also be sufficient for a restructuring like this. Some arguments against include:
- All existing users (especially beginners who rely either on memory when navigating ("The option I want can be activated by pressing down arrow twice, then right arrow, and then down arrow three times", or on menu accelerator keys) would be confused. Note that even small changes like change of accelerators in the tools menu, or move of sound card from one panel to another cause a lot of criticism against NVDA (which is in my view at least partially warranted, since these changes are mostly not done on user requests)
- The current structure reflects quite closely the structure in JAWS, so for people coming from it to NVDA it is mostly familiar (discounting the different way of managing application specific settings)
- From my, admittedly short (around 8 month 5 years ago, or so) time teaching screen readers to beginner users, for them function like config profiles, input gestures or dictionaries are too advanced, and they have no use for them. It is hard as it is for them to learn how to use computer with a screen reader, so as long as they can change basic things from the settings panel they're fine.

> Anyway, one interesting point in the current proposal (and which could help understanding NVDA configuration) is the idea to gather all elements related to configuration in a specific sub-menu. 

I don't know how the current menu structure came to life, but I always thought that items used to revert to saved config and restore it to factory defaults were placed at the top level of the menu on purpose. That makes it possible to ask someone to fix NVDA broken by the unwanted config change, and therefore these items should remain at the visible place in the main menu structure.
Finally I agree with @CyrilleB79  that accelerator conflicts are quite rare, and we  should fix them as they arise, and not restructure the menu because some conflict may occur in the future. Also conflicts in the main and preferences menu didn't occur in the past as far as I can remember.

**Comment by Adriani90** [comment]
> All existing users (especially beginners who rely either on memory when navigating ("The option I want can be activated by pressing down arrow twice, then right arrow, and then down arrow three times", or on menu accelerator keys) would be confused.

I agree with you on this point in a short term, however I think a more structured approach will make things more clear especially for new users in the mid to long term and in the end everyone will be able to adapt.

> Note that even small changes like change of accelerators in the tools menu, or move of sound card from one panel to another cause a lot of criticism against NVDA (which is in my view at least partially warranted, since these changes are mostly not done on user requests)

I think software developers will never see a world where this does not happen. Some users will always firstly critisize designs which get changed before accepting them. This was also the case when NVDA's settings dialogs changed in 2018 and this happened also when Windows evolved in different stages, even when NVDA changed the python version there were a lot of worries which in the end did not realize.
In the end we will always have to take the risk that some people do not like some changes, especially when it comes to UX design.

Thank you all for this valuable discussion. Given the limited support for this change I guess it makes indeed sense to close this for now unless there are other strong arguments for this change. If there are any, we can rethink this and reopen at a later stage.

**Comment by lukaszgo1** [comment]
> I think software developers will never see a world where this does not happen. Some users will always firstly critisize designs which get changed before accepting them. 

They surely do. However when the change does not offer anything which personally improves software for them and forces a change of behavior, I believe they have a right to complain, especially since there always be much less developers who did the change then they are users.


> This was also the case when NVDA's settings dialogs changed in 2018

Note that this enhancement was explicitly requested in various tickets by at least several users, so it was partially user driven.

> In the end we will always have to take the risk that some people do not like some changes, especially when it comes to UX design.

That is true, but we certainly can, as you did here, start by discussing about ratio of benefit to annoyance for users before the change.
